slither/tools/read_storage/read_storage.py

```diff
@@ -234,22 +364,143 @@ def get_slot_values(self, slot_info: SlotInfo) -> None:
+                    return sv
+        return None
+
+    @staticmethod
+    def find_hardcoded_slot_in_asm_str(
+        inline_asm: str, contract: Contract
+    ) -> Optional[StateVariable]:
+        """
+        Searches a block of assembly code (given as a string) for a sload from a literal storage slot.
+        Does not work if the argument passed to sload does not start with "0x", i.e., `sload(add(1,1))`
+        or `and(sload(0), 0xffffffffffffffffffffffffffffffffffffffff)`.
+
+        Args:
+            inline_asm: a string containing all the code in an assembly node (node.inline_asm for solc < 0.6.0).
+
+        Returns:
+            A newly created StateVariable representing the Literal bytes32 slot, if one is found, otherwise None.
+        """
+        asm_split = inline_asm.split("\n")
+        for asm in asm_split:
+            if "sload(" in asm:  # Only handle literals
+                arg = asm.split("sload(")[1].split(")")[0]
+                if arg.startswith("0x"):
+                    exp = Literal(arg, ElementaryType("bytes32"))
+                    sv = StateVariable()
+                    sv.name = "fallback_sload_hardcoded"
+                    sv.expression = exp
+                    sv.is_constant = True
+                    sv.type = exp.type
+                    sv.set_contract(contract)
+                    return sv
+        return None
+
+    def find_hardcoded_slot_in_exp(
+        self, exp: "Expression", contract: Contract
+    ) -> Optional[StateVariable]:
+        """
+        Parses an expression to see if it contains a sload from a literal storage slot,
+        unrolling nested expressions if necessary to determine which slot it loads from.
+        Args:
+            exp: an Expression object to search.
+            contract: the Contract containing exp.
+
+        Returns:
+            A newly created StateVariable representing the Literal bytes32 slot, if one is found, otherwise None.
+        """
+        if isinstance(exp, AssignmentOperation):
+            exp = exp.expression_right
+        while isinstance(exp, BinaryOperation):
+            exp = next(
+                (e for e in exp.expressions if isinstance(e, (CallExpression, BinaryOperation))),
+                exp.expression_left,
             )
+        while isinstance(exp, CallExpression) and len(exp.arguments) > 0:
+            called = exp.called
+            exp = exp.arguments[0]
+            if "sload" in str(called):
+                break
+        if not isinstance(exp, Literal) and isinstance(
+            exp,
+            (BinaryOperation, UnaryOperation, Identifier, TupleExpression, TypeConversion),
+        ):
+            exp = ConstantFolding(exp, "bytes32").result()
+        if (
+            isinstance(exp, Literal)
+            and isinstance(exp.type, ElementaryType)
+            and exp.type.name in ["bytes32", "uint256"]
+            # and (str(exp.value).startswith("0x") or isinstance(exp.value, bytes))
```

Comment:
Is this needed or can be removed?

Reply:
I think it's needed. This is basically the definition of the hardcoded slot that `find_hardcoded_slot_in_exp` is looking for.

I know ConstantFolding always returns a Literal (if it doesn't raise a NotConstant exception), but there's no guarantee that ConstantFolding is used in the lines above this, hence the check for `isinstance(exp, Literal)`. Perhaps I could get rid of the `isinstance(exp.type, ElementaryType)`, since it appears that will always be the case when ConstantFolding is used. But then I think I need the `exp.type.name in ["bytes32", "uint256"]` because I'm not interested in constant addresses, strings or bools.

Does that make sense?